addon/-private/utils/model-definition.ts

```diff
@@ -4,63 +4,59 @@ import {
   RelationshipDefinition,
   ModelDefinition
 } from '@orbit/data';
+import { clone, deepMerge } from '@orbit/utils';
 
-import Model from '../model';
-
-const MODEL_DEFINITIONS = Symbol('@orbit:modelDefinitions');
-const PROPERTY_NOTIFIERS = Symbol('@orbit:poropertyNotifiers');
-
-export type Notifier = (record: Model) => void;
+const MODEL_DEFINITION = Symbol('@orbit:modelDefinition');
 
 export function getModelDefinition(proto: object): ModelDefinition {
-  if (proto[MODEL_DEFINITIONS]) {
-    return proto[MODEL_DEFINITIONS] as ModelDefinition;
+  if (proto[MODEL_DEFINITION]) {
+    return proto[MODEL_DEFINITION] as ModelDefinition;
   } else {
-    proto[MODEL_DEFINITIONS] = {};
-    return proto[MODEL_DEFINITIONS];
+    proto[MODEL_DEFINITION] = {};
+    return proto[MODEL_DEFINITION];
   }
 }
 
-export function getPropertyNotifiers(proto: object): Record<string, Notifier> {
-  proto[PROPERTY_NOTIFIERS] = (proto[PROPERTY_NOTIFIERS] || {}) as Record<
-    string,
-    Notifier
-  >;
-  return proto[PROPERTY_NOTIFIERS];
+export function extendModelDefinition(
+  proto: object,
+  modelDefinition: ModelDefinition
+): void {
+  if (proto[MODEL_DEFINITION]) {
+    proto[MODEL_DEFINITION] = deepMerge(
+      clone(proto[MODEL_DEFINITION] as ModelDefinition),
+      modelDefinition
+    );
+  } else {
+    proto[MODEL_DEFINITION] = modelDefinition;
+  }
 }
```

Comment:
I've found that we can greatly minimize the number of clone operations if we refactor this to include another property, `proto[MODEL_DEFINITION_FOR]`, that tracks the current `proto` that the model def applies to:

```ts
const MODEL_DEFINITION_FOR = Symbol('@orbit:modelDefinitionFor');

export function extendModelDefinition(
  proto: object,
  modelDefinition: ModelDefinition
): void {
  if (proto[MODEL_DEFINITION] && proto[MODEL_DEFINITION_FOR]) {
    let currentDef = proto[MODEL_DEFINITION];
    if (proto[MODEL_DEFINITION_FOR] !== proto) {
      currentDef = clone(currentDef);
      proto[MODEL_DEFINITION_FOR] = proto;
    }
    proto[MODEL_DEFINITION] = deepMerge(currentDef, modelDefinition);
  } else {
    proto[MODEL_DEFINITION] = modelDefinition;
    proto[MODEL_DEFINITION_FOR] = proto;
  }
}
```

This reduces the number of clone ops to the number of models that are extended (in the whole test suite this reduces the clone count from 40 -> 7).